### PR TITLE
Functionality for multiple tasks per phase

### PR DESCRIPTION
**Pull request by jimmykodes**
This will duplicate the submission object as many times as necessary to have a submission for each task in a phase, all pointing to the same data file. Since leaderboards are per submission, I'm not sure how users would get scores for all the submissions? but they all run now. Can test w/ the attached bundle (submission is in bundle).
[Bundle.zip](https://github.com/codalab/competitions-v2/files/3254529/Bundle.zip)

### CODE REVIEW

File: src/apps/api/serializers/submissions.py [hunk 1]
```diff
@@ -89,17 +90,30 @@ def create(self, validated_data):
...
         print("Updated to...")
         print(validated_data)
 
         if validated_data["status"] == Submission.SCORING:
             # Start scoring because we're "SCORING" status now from compute worker
             from competitions.tasks import run_submission
-            run_submission(instance.pk, is_scoring=True)
+            task_id = validated_data.get('task_pk')
+            if not task_id:
+                return Response(
```
Comment (ckcollab) [inline src/apps/api/serializers/submissions.py]: Raise `ValidationError`

File: src/apps/competitions/models.py [hunk 2]
```diff
@@ -194,7 +194,13 @@ def save(self, ignore_submission_limit=False, **kwargs):
 
     def start(self):
         from .tasks import run_submission
-        run_submission.apply_async((self.pk,))
+        for i, task in enumerate(self.phase.tasks.all()):
```
Comment (ckcollab) [inline src/apps/competitions/models.py]: imo pull this out

File: src/apps/api/tests/test_submission_scores.py [hunk 3]
```diff
@@ -19,17 +18,22 @@ def setUp(self):
         self.leaderboard = LeaderboardFactory(competition=self.competition)
         self.column = ColumnFactory(leaderboard=self.leaderboard, key='test')
         self.submission = self.make_submission()
-        self.client = APIClient()
 
-    def make_submission(self, **kwargs):
+    def make_submission(self, create_score=True, **kwargs):
         kwargs.setdefault('owner', self.user)
         kwargs.setdefault('phase', self.phase)
+        parent_sub = kwargs.pop('parent_submission') if 'parent_submission' in kwargs else None
```
Comment (ckcollab) [inline src/apps/api/tests/test_submission_scores.py]: Can do this instead, I believe..
```python
kwargs.pop('parent_submission', None)
```

File: src/apps/competitions/models.py [hunk 4]
```diff
@@ -322,6 +326,12 @@ def cancel(self):
             return True
         return False
 
+    def check_children(self):
```
Comment (ckcollab) [inline src/apps/competitions/models.py]: I'm not a huge fan of this name, something like `check_children_completed` is _maybe_ better.. hmm..

File: src/apps/competitions/tasks.py [hunk 5]
```diff
@@ -94,7 +94,32 @@
...
+
+    run_args['program_data'] = make_url_sassy(
+        path=submission.data.data_file.name if not is_scoring else task.scoring_program.data_file.name
+    )
+    run_args['result'] = make_url_sassy(
+        path=submission.result.name,
+        permission='w' if not is_scoring else 'r'
+    )
+    run_args['task_pk'] = task.id
+    detail_names = SubmissionDetails.DETAILED_OUTPUT_NAMES_PREDICTION if not is_scoring else SubmissionDetails.DETAILED_OUTPUT_NAMES_SCORING
```
Comment (ckcollab) [inline src/apps/competitions/tasks.py]: make multiline?

File: src/apps/competitions/tasks.py [hunk 6]
```diff
@@ -94,7 +94,32 @@
 ]
 
 
-def send(submission, run_args):
+def send(submission, task, is_scoring, run_args):
```
Comment (ckcollab) [inline src/apps/competitions/tasks.py]: should maybe rename this function as well.. hm..

File: src/apps/competitions/tests/test_phase_migration.py [hunk 7]
```diff
@@ -95,3 +93,9 @@ def test_all_submissions_migrated_before_changing_phase_status(self):
 
         mock_start = self.mock_migration()
         assert mock_start.call_count == 0
+
+    def test_only_parent_submissions_migrated(self):
+        self.parent_submission = self.make_submission()
```
Comment (ckcollab) [inline src/apps/competitions/tests/test_phase_migration.py]: can we assert in here that many children exist to make sure we dont have just a parent and no children so we get a false positive?

File: src/apps/competitions/tests/test_submissions.py [hunk 8]
```diff
@@ -72,27 +69,17 @@ def test_max_submissions_not_counting_other_user_submissions(self):
...
-            try:
-                Submission.objects.get(name='Find Me')
-                assert False, "Submission should not have been created"
-            except Submission.DoesNotExist:
-                pass
+        self.assertRaises(PermissionError, self.make_submission)
+        assert not Submission.objects.filter(name='Find Me').exists()
 
+    def test_children_submissions_dont_count_toward_max(self):
+        self.make_submission(parent=self.make_submission())  # Should only count as one submission
```
Comment (ckcollab) [inline src/apps/competitions/tests/test_submissions.py]: in general, imo, comments should be above the code and not to the right, but it definitely fits in some cases -- seems a bit much here

File: src/factories.py [hunk 9]
```diff
@@ -139,6 +139,16 @@ class SubmissionScoreFactory(DjangoModelFactory):
     class Meta:
         model = SubmissionScore
 
-    submission = factory.SubFactory(SubmissionFactory)
     column = factory.SubFactory(ColumnFactory)
     score = factory.LazyAttribute(lambda n: random.choice(range(1, 11)) / 10)
+
+    @post_generation
+    def submissions(self, created, extracted, **kwargs):
+        if not created:
```
Comment (ckcollab) [inline src/factories.py]: maybe add a comment for some of this magic? what is extracted?

File: src/static/riot/competitions/detail/submission_manager.tag [hunk 10]
```diff
@@ -79,14 +79,36 @@
...
         <div class="content">
-            <submission-modal></submission-modal>
+            <div if="{!!selected_submission && !_.get(selected_submission, 'has_children', false)}">
+                <submission-modal submission="{selected_submission}"></submission-modal>
+            </div>
+            <div if="{!!selected_submission && _.get(selected_submission, 'has_children', false)}">
+                <div class="ui large green pointing menu">
+                    <div each="{child, i in _.get(selected_submission, 'children')}"
+                         class="parent-modal item"
+                         data-tab="{admin_: is_admin()}child_{i}">
```
Comment (ckcollab) [inline src/static/riot/competitions/detail/submission_manager.tag]: whoa that's fancy

File: src/static/riot/competitions/detail/submission_manager.tag [hunk 11]
```diff
@@ -197,12 +229,54 @@
...
+                    return score.column_key === column.key
+                })[0]
+                return [score.score, score.id]
+            } catch {
+                return ['', '']
+            }
+        }
+
+        self.submission_clicked = function (submission) {
+            submission = _.defaultsDeep({}, submission) // stupid workaround to not modify the original submission object
```
Comment (ckcollab) [inline src/static/riot/competitions/detail/submission_manager.tag]: comments above the line, typically

File: src/static/riot/competitions/detail/submission_manager.tag [hunk 12]
```diff
@@ -197,12 +229,54 @@
...
                 submission.admin = true
             }
-            CODALAB.events.trigger('submission_clicked', submission)
-            $(self.refs.modal).modal('show');
+            self.selected_submission = submission
+            self.update()
+            $(self.refs.modal)
+                .modal({
+                    onShow: () => {
+                        if(_.get(self.selected_submission, 'has_children', false)){
```
Comment (ckcollab) [inline src/static/riot/competitions/detail/submission_manager.tag]: quick comment here?

**Comment by jimmykodes** [comment]
@ckcollab this is waiting on your hand tests, correct?

**Comment by ckcollab** [comment]
Should we make a separate issue for displaying output from multiple tasks at once?

**Comment by jimmykodes** [comment]
:shipit:

**Pull request merged**